Add `gh issue develop` command

Bestra opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

This PR introduces a new `develop` subcommand for `gh issue`.  It has two main functions for now. I think the usage block is a decent summary:
``` go
		Use:   "develop [flags] {<number> | <url>}",
		Short: "Manage linked branches for an issue",
		Example: heredoc.Doc(`
			$ gh issue develop --list 123 # list branches for issue 123
			$ gh issue develop --list --issue-repo "github/cli" 123 list branches for issue 123 in repo "github/cli"
			$ gh issue develop --list https://github.com/github/cli/issues/123 # list branches for issue 123 in repo "github/cli"
			$ gh issue develop 123 --name "my-branch" --base main
			$ gh issue develop 123 --checkout # checkout the branch for issue 123 after creating it
			`),
```
Users can provide an issue number or a URL.  They can also specify the issue's repo separately via the `issue-repo` flag. If the flag specifies a different repo than the provided URL we'll return an error.

### Feature detection

This command currently relies on a feature-flagged part of our GraphQL schema. We make an introspection query to see if the `LinkedBranch` type exists, and bail if it doesn't:
```shell
  GH_HOST=github.localhost  gh issue develop http://github.localhost/monalisa/smile/issues/1 --list
⣽the `gh issue develop` command is not currently available
```
### Listing existing linked branches with `--list`

Given I have an issue with some linked branches
<img width="407" alt="CleanShot 2022-09-14 at 15 08 14@2x" src="https://user-images.githubusercontent.com/2043348/190241524-bf144b7e-eff3-4d0d-90f4-b88e3c1f8807.png">

Passing the `--list` flag will spit out the first 30 linked branches.  For now I've kept pagination out of scope but I don't think it'd be much work to add.
<img width="837" alt="CleanShot 2022-09-26 at 14 38 44@2x" src="https://user-images.githubusercontent.com/2043348/192354853-f814a051-4b7c-4ef3-a88e-725fef0082d7.png">


### Creating linked branches (omitting `--list`)

Without the `--list` flag this command will create a new linked branch in the base repo. 
Creating a linked branch and checking it out locally:

https://user-images.githubusercontent.com/2043348/190250169-2b1b969b-e2a1-4fc4-847a-ecbf602ded3b.mp4

## Notables

The GraphQL endpoints for these operations are currently feature flagged. We're the first real integrators.  We don't yet have an endpoint that can link existing branches to issues. I'd like to do that as a followup to this initial PR if possible. Once that's done I can add `--link` and `--unlink` flags, as we do have an endpoint that can delete an existing link (without removing the underlying branch).  I think the best coarse of action 

Ideally I think this command would parallel the `Development` menu on the issue sidebar as closely as possible, which would include showing linked *PRs* as well as linked branches, but for now that's firmly out of scope.

### Hello golang!

I've tried to avoid as much novelty as possible in writing this new command, taking existing functionality and matching the style of the codebase. That said it's my first time writing go in anger and I'd be very happy to have a quick pairing session if it feels like reviewing this async is going to be difficult.

Bestra committed:
WIP

Bestra committed:
WIP

Bestra committed:
WIP

Bestra committed:
Add flag to features list

Bestra committed:
WIP

Bestra committed:
FIll in all the missing stuff in the test setup

Bestra committed:
Get that test mostly working

Bestra committed:
Make sure struct fields are public

Bestra committed:
Use GraphQLQuery test helper

GraphQLMutation expects a single `input` variable

Bestra committed:
WIP add --checkout option

Bestra committed:
List linked branches works

Bestra committed:
Use shared fn to look up issue and its repo

Bestra committed:
Add tests for checkout flag

Bestra committed:
WIP tests for --list flag

Bestra committed:
Error when issue-repo doesn't match issue url

Bestra committed:
Omit name param from mutation when blank

As it currently sits the backend won't ignore the param. I'll be looking to
add a PR for this but for now we'll remove it.

Bestra committed:
self-review cleanup

Bestra committed:
Merge branch 'trunk' into create-linked-branch-command

Bestra commented:
I'm in the process of
1. fixing those failing tests
2. modifying the code so we properly omit the `name` argument from the `CreateLinkedBranch` mutation if it's blank
3. Doing feature detection (by way of schema introspection) so that we can throw an error if the endpoint isn't available (really this is for GHES)
4. Adding a prompt footer to the `--list` command telling users how to check out one of those branches

Bestra committed:
Properly check for omitted name param, fix stubs

Bestra commented on api/queries_pr.go:
```diff
@@ -304,10 +304,7 @@ func CreatePullRequest(client *Client, repo *Repository, params map[string]inter
-		"input": inputParams,
-	}
-
+	variables := inputParams
```
whoops

Bestra committed:
Add feature detection query

Bestra committed:
Revert accidental `queries_pr` change

Bestra committed:
Add tests, update stubs for feature detection

Bestra commented:
@vilmibm one edge case (kind of but not really) that I've just now realized is that in the issue develop menu on dotcom we show the branch name and the repo NWO
<img width="351" alt="CleanShot 2022-09-16 at 13 50 32@2x" src="https://user-images.githubusercontent.com/2043348/190701734-3d4501cd-9e97-4ef4-a527-f386a80d1a7f.png">

we should probably surface the NWO along with the branch name in the list results too.
```
Showing linked branches for monalisa/test#1

"my-branch" in http://github.localhost/monalisa/smile
"some-other-branch" in http://github.localhost/monalisa/test
"so-many-branches-why" in http://github.localhost/monalisa/smile
```
feels...well like something. I think having the GitHub url in there makes sense in some capacity

vilmibm commented:
I think surfacing the NWO makes sense, but maybe use the table printer so it looks more like:

```
my-branch            http://github.localhost/monalisa/smile
some-other-branch    http://github.localhost/monalisa/test
so-many-branches-why http://github.localhost/monalisa/smile
```

using cyan for branch name as that's a pretty consistent scheme that we have. 

I don't think these would ever be cross host, right? so just doing NWO is probably fine. I don't think we can say with confidence that after running this command a user would want to click the repo link, so it being a link feels like some extra noise here.

Bestra commented:
> I don't think these would ever be cross host, right?

No, I think these will always be on the same host.

I'm at a team summit this week but I'll ping you for another review once I've used the table printer to add the urls.  Thanks for the advice!

Bestra committed:
Show repo url for linked branches in tty

Bestra commented:
@vilmibm That table printer is great! We now print out the branch url along with its name 
<img width="837" alt="CleanShot 2022-09-26 at 14 38 44@2x" src="https://user-images.githubusercontent.com/2043348/192354924-58b537c6-1042-41b4-a454-4af78cc14a91.png">

I've also added in some feature detection via schema introspection (hey it rhymes) in https://github.com/cli/cli/pull/6254/commits/76d975524e02b723dec19620f3a973df913ae309. I _think_ this is ready for another look on your part.

Bestra commented:
I have a [PR open to remove ](https://github.com/github/github/pull/240028)the `branch_for_issue_api` feature flag. Depending on timing I can remove the flag from this PR first, or I can come back to remove it after we merge.

Bestra commented on api/client.go:
```diff
@@ -19,7 +19,7 @@ const (
 	cacheTTL        = "X-GH-CACHE-TTL"
 	graphqlFeatures = "GraphQL-Features"
-	mergeQueue      = "merge_queue"
+	features        = "merge_queue,branch_for_issue_api"
```
The fields and mutations for this command are now public https://docs.github.com/en/graphql/reference/mutations#createlinkedbranch so I think we can remove the flag here 
```suggestion
	features        = "merge_queue"
```

Bestra committed:
Remove released feature flag

`branch_for_issue_api` is no longer in the dotcom codebase so we can remove it from here

Bestra commented:
@vilmibm do you think we'll be able to get this one reviewed before the next cycle? Apologies for not keeping up, it's an "out of sight out of mind" kind of thing now that I'm working in other areas 😄 . Let me know if I can make the review easier on you, and thanks again!

vilmibm reviewed (approved):
yesssssssssssssSSS

vilmibm commented:
@Bestra yes!!! merging it today :D thanks again for your thorough work, patience, and kindness!

Bestra commented:
:steve-holt-gesture:

Hilljackcantspell commented:
This is why I dont help much Kindness always mistake for ignorance Have a nice day

Hilljackcantspell commented:
S O I
